CoreObjectModel/MutableMetadataModel/Copier.cs

```diff
@@ -2404,7 +2404,17 @@ class MetadataDispatcher : MetadataVisitor {
         if (nestedTypeReference is Dummy) return nestedTypeReference;
         object copy;
         if (keepAsDefinition && this.DefinitionCache.TryGetValue(nestedTypeReference, out copy)) return (INestedTypeReference)copy;
-        if (this.ReferenceCache.TryGetValue(nestedTypeReference, out copy)) return (NestedTypeReference)copy;
+        if (this.ReferenceCache.TryGetValue(nestedTypeReference, out copy))
+        {
+            NestedTypeReference cachedCopy = (NestedTypeReference)copy;
+            // There are cases where the IsValueType flag changes on the same object reference
+            // and if that happens it means we have a reference that requires that bit and
+            // so we need to update our cached value to reflect that
+            // See PEFileToObjectModel.GetTypeRefReferenceAtRow for reference at:
+            // Line: if (mustBeStruct && typeRefReference != null) typeRefReference.isValueType = true;
+            if (nestedTypeReference.IsValueType) cachedCopy.IsValueType = true;
```

Comment:
Unfortunately, ITypeReference.IsValueType is problematic in both cases. Some references, such as signatures, include whether the referenced type is a value type while other references don't. Both the original version and this one mean that IsValueType can vary; this just changes when. I might lean toward skipping this part of the change since anything that depends on ITypeReference.IsValueType is inherently fragile (and if you added this to fix a problem in your usage, you likely have a bug).

Reply:
Yes, I totally agree that this is a bug in the spec (or maybe just the implementation) of .NET. But I think this is a really necessary change. The information that CCI has about the type can change depending on whether you visit a reference that contains the information about it being a value type. If that information is not reflected in the object model then incorrect IL can be generated. The only true way to fix this is by resolving all type references and that is totally unacceptable given CCI's architecture. I'm happy to talk more about this if you'd like.